crates/biome_js_analyze/src/lint/nursery/use_import_extensions.rs

```diff
@@ -0,0 +1,201 @@
+    /// ```js,expect_diagnostic
+    /// import "../.";
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// import "biome";
+    /// ```
+    /// ```js
+    /// import "./foo.js";
+    /// ```
+    /// ```js
+    /// import "./bar/index.js";
+    /// ```
+    ///
+    /// ## Caveats
+    ///
+    /// Rule does not yet check filesystem for file type. It tries to guess which extension
+    /// it should add based on the file extension of the current file and the import path.
+    /// When applying the suggested fix, make sure to verify that the file type is correct.
+    ///
+    pub UseImportExtensions {
+        version: "next",
+        name: "useImportExtensions",
+        recommended: false,
+    }
+}
+
+impl Rule for UseImportExtensions {
+    type Query = Ast<JsModuleSource>;
+    type State = UseImportExtensionsState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let binding = ctx.query();
+
+        let Ok(module_path) = binding.inner_string_text() else {
+            return None;
+        };
+
+        let file_ext = ctx.file_path().extension()?.to_str()?;
+
+        get_extensionless_import(file_ext, &module_path)
+    }
+
+    fn diagnostic(ctx: &RuleContext<Self>, _: &Self::State) -> Option<RuleDiagnostic> {
+        let node = ctx.query();
+
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                node.range(),
+                markup! {
+                    "Import extension is required for relative imports."
+                },
+            )
+            .note(markup! {
+                "Explicit import extension makes tooling faster and is more consistent."
+            }),
+        )
+    }
+
+    fn action(ctx: &RuleContext<Self>, state: &Self::State) -> Option<JsRuleAction> {
+        let mut mutation = ctx.root().begin();
+        let node = ctx.query();
+
+        mutation.replace_node(
+            node.clone(),
+            make::js_module_source(make::js_string_literal(&state.suggestion)),
+        );
+
+        Some(JsRuleAction {
+            category: ActionCategory::QuickFix,
+            applicability: Applicability::MaybeIncorrect,
+            message: markup! {
+                "Add potential import extension "<Emphasis>"."{state.potential_import_ext}</Emphasis>"."
+            }.to_owned(),
+            mutation,
+        })
+    }
+}
+
+pub struct UseImportExtensionsState {
+    potential_import_ext: String,
+    suggestion: String,
+}
+
+const GUARANTEED_INDEX_FILE_ENDINGS: &[&str] = &["/", "..", "."];
+
+fn get_extensionless_import(
+    file_ext: &str,
+    module_path: &TokenText,
+) -> Option<UseImportExtensionsState> {
+    if !module_path.starts_with('.') {
+        return None;
+    }
+
+    if Path::new(module_path.text()).extension().is_some() {
+        return None;
+    }
+
+    let mut path_parts: Vec<_> = module_path.text().split('/').collect();
```

Comment:
I think we can avoid allocating a vector, and just using the Split iterator.

You can call `peekable()` to check items without consuming the iterator, and `next_back` to pop the items.

You won't need to use the len check anymore 

https://doc.rust-lang.org/std/str/struct.Split.html

Reply:
Or even better, you can use [path::components()](https://doc.rust-lang.org/std/path/struct.Path.html#method.components)

Reply:
I updated implementation to use PathBuf, not sure if it is faster or not compared to using a vector. If not how should I modify path?

Reply:
Windows tests fail because it uses OsPath and in windows path slash is `\`.
```
 11 │ + import·"./../index.js"
 11 │ + import·"./..\index.js"
```
Ideas?